src/policy/space.rs

```diff
@@ -46,8 +46,42 @@ pub trait Space<VM: VMBinding>: 'static + SFT + Sync + Downcast {
     /// Currently after we create a boxed plan, spaces in the plan have a non-moving address.
     fn initialize_sft(&self);
 
+    /// A check for the obvious out-of-memory case: if the requested size is larger than
+    /// the heap size, it is definitely an OOM. We would like to identify that, and
+    /// allows the binding to deal with OOM. Without this check, we will attempt
+    /// to allocate from the page resource. If the requested size is unrealistically large
+    /// (such as `usize::MAX`), it breaks the assumptions of our implementation of
+    /// page resource, vm map, etc. This check prevents that, and allows us to
+    /// handle the OOM case.
+    /// Each allocator that may request an arbitrary size should call this method before
+    /// acquring memory from the space. For example, bump pointer allocator and large object
+    /// allocator need to call this method. On the other hand, allocators that only allocate
+    /// memory in fixed size blocks do not need to call this method.
+    fn oom_size_check(&self, tls: VMThread, size: usize) {
+        let max_pages = self.get_gc_trigger().policy.get_max_heap_size_in_pages();
+        let requested_pages = size >> LOG_BYTES_IN_PAGE;
+        if requested_pages > max_pages {
+            VM::VMCollection::out_of_memory(
+                tls,
+                crate::util::alloc::AllocationError::HeapOutOfMemory,
+            );
+            panic!(
```

Comment:
If `panic!()` immediately follows `Collection::out_of_memory()`, it implies the only way not to reach this `panic!` is doing stack unwinding (or swap-stack or other advanced stack-manipulation techniques) within `out_of_memory`.  I don't know how a VM can do this reliably because unwinding through Rust frames is ABI-specific.  But given that existing code (such as `util::memory::handle_mmap_error`) also assumes this function never returns, we may leave this to another day.

Reply:
We can just return `Address::ZERO` here without panic (maybe with a `warn!`), in the same way as other `out_of_memory()` uses.

Reply:
I changed the method name to `will_go_oom_on_acquire`, and returns a boolean. If the return value is true, the allocator will return `Address::ZERO`.

Reply:
May I suggest `will_oom_on_acquire`?

Reply:
Sounds good. Done.